e2e_playwright/shared/performance.py

```diff
@@ -0,0 +1,136 @@
+    }
+}
+
+new PerformanceObserver(handleEntries).observe({
+    entryTypes: ['longtask', 'measure', 'mark', 'navigation', 'paint', 'long-animation-frame'],
+});
+"""
+
+GET_CAPTURED_TRACES_SCRIPT = """
+window.__stCapturedTraces.profiles = {};
+
+for (const [key, value] of Object.entries(window.__streamlit_profiles__ || {})) {
+    window.__stCapturedTraces.profiles[key] = {
+        entries: value.buffer.filter(Boolean),
+        totalWrittenEntries: value.totalWrittenEntries,
+    };
+}
+
+JSON.stringify(window.__stCapturedTraces)
+"""
+
+
+def is_supported_browser(page: Page) -> bool:
+    browser = page.context.browser
+    browser_name = browser.browser_type.name if browser is not None else "unknown"
+    # Only measure performance for Chromium browsers since it relies on
+    # Chrome DevTools Protocol under the hood
+    return browser_name == "chromium"
+
+
+def start_capture_traces(page: Page):
+    """
+    Start capturing traces using the PerformanceObserver API.
+    """
+    if is_supported_browser(page):
+        page.evaluate(CAPTURE_TRACES_SCRIPT)
+
+
+@contextmanager
+def with_cdp_session(page: Page):
+    """
+    Create a new Chrome DevTools Protocol session.
+    Detach the session when the context manager exits.
+    """
+    if not is_supported_browser(page):
+        raise RuntimeError(
+            "Chrome DevTools Protocol is only supported on Chromium-based browsers."
+        )
+
+    client = page.context.new_cdp_session(page)
+    yield client
+    client.detach()
+
+
+@contextmanager
+def measure_performance(
+    page: Page, *, test_name: str, cpu_throttling_rate: int | None = None
+):
+    """
+    Measure the performance of the page using the native performance API from
+    Chrome DevTools Protocol.
+    @see https://github.com/puppeteer/puppeteer/blob/main/docs/api/puppeteer.page.metrics.md
+
+    Args:
+        page (Page): The page to measure performance on.
+        cpu_throttling_rate (int | None, optional): Throttling rate as a slowdown factor (1 is no throttle, 2 is 2x slowdown, etc). Defaults to None.
+    """
+    with with_cdp_session(page) as client:
+        if cpu_throttling_rate is not None:
+            client.send("Emulation.setCPUThrottlingRate", {"rate": cpu_throttling_rate})
+
+        client.send("Performance.enable")
+
+        # Run the test
+        yield
+
+        metrics_response = client.send("Performance.getMetrics")
+        captured_traces_result = client.send(
+            "Runtime.evaluate",
+            {"expression": GET_CAPTURED_TRACES_SCRIPT},
+        )["result"]
+        captured_traces = captured_traces_result.get("value", "{}")
+        parsed_captured_traces = json.loads(captured_traces)
+
+        # Ensure the directory exists
+        os.makedirs("./performance-results", exist_ok=True)
+
+        timestamp = datetime.datetime.now().strftime("%Y%m%d%H%M%S")
+
+        with open(f"./performance-results/{timestamp}_{test_name}.json", "w") as f:
```

Comment:
Thanks for the comment @Asaurus1 ! As written, this follows our existing paradigms for other artifacts such as `test-results` and `__snapshots__`. With that said, I agree that we can improve it! Here's a follow-up PR to make this better for all artifacts from tests that live in `e2e_playwright`: https://github.com/streamlit/streamlit/pull/10039

Reply:
I'm... not sure that's entirely true. Anecdotally I've never had an issue with either the snapshots or test-results.

`test-results` for example... as best I can tell the path is defined here: https://github.com/streamlit/streamlit/blob/20502d2ee73635081f01e4717f1747bb35ffa7a6/Makefile#L345. That `--output` argument is only provided if you hook `pytest` using `make playwright`, at which point the CD will _always_ be `e2e_playwright` because of the `cd` a few lines above. If you run pytest using a different hook, you either have to provide the `--ouput` argument yourself or just not provide it (in which case, there's no output).

With `__snapshots__` it's a bit different; you'll get a `FileNotFound` error if you run the test from the wrong directory because of [these lines here](https://github.com/streamlit/streamlit/blob/20502d2ee73635081f01e4717f1747bb35ffa7a6/e2e_playwright/conftest.py#L581-L586), but running the tests doesn't actually generate the snapshots. Snapshot generation happens in [this scriptlet](https://github.com/streamlit/streamlit/blob/20502d2ee73635081f01e4717f1747bb35ffa7a6/scripts/update_e2e_snapshots.py#L35), and it _does_ actually have CWD-less logic to find the project root and locate the shapshots in the e2e directory regardless of where you are when you run it. Like `test-results` it's also not hooked in if you use an entry point other than the `Makefile`.

What makes the `performance-results` situation unique is that the output of the files is A) it's on by default -- you don't need to pass a `--output-performance-results=./performance-results` argument to `pytest`; and B) it outputs files every test -- unlike `__snapshots__` which only _read_ files at test-time.

Reply:
Yes, that's correct that commands in the makefile are running within the context of the "correct" directory hence why running `make` commands won't run into this issue.

However, there are many other valid workflows and thus this issue of creating 3 directories in the wrong location does show up. For example, this issue crops up by running pytest in a directory that is not `e2e_snapshots`. Running `pytest ./e2e_playwright/st_button_test.py` from the git root directory will create all 3 of those directories (`__snapshots__`, `test-results`, and `performance-results`).

Again, thank you for your comments, I have a fix for all of these in https://github.com/streamlit/streamlit/pull/10039